### PR TITLE
net: http: Fix HTTP_DATA_FINAL notification

### LINKED ISSUE #38428: http_client response callback always reports final_data == HTTP_DATA_FINAL
**Describe the bug**
The response callback triggered by http_client_req reported always HTTP_DATA_FINAL, although more data was left and also more calls to the callback function occurred.

**To Reproduce**
1. setup an instance of struct http_request with a buffer size of 256 bytees
2. execute http_client_req which will result in a response of 784 bytes
3. -> multiple response callbacks with HTTP_DATA_FINAL, actually every single one

**Expected behavior**
The flag final_data should be set to HTTP_DATA_FINAL only for the last call to the callback function.

**Impact**
It is more an annoyance, as the data_len field can be checked for 0. Which seems to be a good indicator for the last packet chunk.

### PR DESCRIPTION
**Pull request by rlubos**
HTTP_DATA_FINAL was incorrectly notified in case Content Length field
was present in the HTTP respone - in such case it was set for every
response fragment, not only the last one.

Fix this by relying on `message_complete` flag instead of
`http_should_keep_alive()` function to determine whether to notify
HTTP_DATA_FINAL or not. As the HTTP parser calls the
`on_message_complete()` callback in either case (response is chunked or
not), this seems to be a more reasonable apporach to determine whether
the fragment is final or not.

Additinally, instead of calling response callback for
`on_body`/`on_message_complete` separately, call it directly from
`http_wait_data()` function, after the parsing round. This fixes the
case when headers were not reported correctly when the provided buffer
was smaller than the total headers length, resulting in corrupted data
being reported to the user.

Fixes #38428

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

### CODE REVIEW

**Comment by rlubos** [comment]
CC @benediktibk

File: subsys/net/lib/http/http_client.c [hunk 1]
```diff
@@ -434,15 +414,32 @@ static int http_wait_data(int sock, struct http_request *req)
...
 				received);
 		}
 
 		total_received += received;
-		offset += received;
 
-		if (offset >= req->internal.response.recv_buf_len) {
-			offset = 0;
+		if (req->internal.response.cb) {
+			if (!req->internal.response.message_complete) {
```
Comment (jukkar) [inline subsys/net/lib/http/http_client.c:424]: IMHO it is always easier to read the code if there is no `!` in it (less mental twisting needed). So would recommend
```
if (req->internal.response.message_complete) {
    DATA_FINAL case
} else {
    DATA_MORE case
}
```

File: subsys/net/lib/http/http_client.c [hunk 2]
```diff
@@ -411,17 +383,25 @@ static void http_client_init_parser(struct http_parser *parser,
...
 {
 	int total_received = 0;
-	size_t offset = 0;
 	int received, ret;
 
 	do {
-		received = zsock_recv(sock, req->internal.response.recv_buf + offset,
-				req->internal.response.recv_buf_len - offset,
-				0);
+		received = zsock_recv(sock, req->internal.response.recv_buf,
```
Comment (jukkar) [inline subsys/net/lib/http/http_client.c:389]: What happens if we receive data one byte at a time, then the earlier data is discarded after each recv call

Comment (rlubos) [reply subsys/net/lib/http/http_client.c:389]: The parser processes the data one byte at a time (https://github.com/zephyrproject-rtos/zephyr/blob/50ff77fa910dcad29167b389a6c7c67ade8f6f17/subsys/net/lib/http/http_parser.c#L784) so it's shouldn't be an issue for the parser.

If we want to restore the "offsetting" functionality to report larger chunks to the application in such case, I can do this, it'd complicate the logic a bit however.

Comment (rlubos) [reply subsys/net/lib/http/http_client.c]: Actually, it wasn't that bad - I've restored the offset and updated the logic - less noise on the application side with chunked responses.

Comment (jukkar) [reply subsys/net/lib/http/http_client.c]: :+1: thanks! I think it is better if application can receive more than one byte at a time.

**Review by jukkar — APPROVED** [review]

**Review by pfalcon — APPROVED** [review]

**Comment by mglettig** [comment]
@rlubos I observed some strange behavior with Zephyr v2.7 today. It looks like sometimes `rsp->body_start` is set in the middle of a response (long after I already received the start of the body). My application was checking this pointer and it took this pointer over the `rsp->recv_buf` pointer if it wasn't set to NULL. For version < Zephyr v2.6 this leaded to correct behavior now it leads to corrupt data. I will now simplify my code and only check `rsp->body_start` add the beginning of a message to figure out when the header is finished. Still I think this is strange behavior and should be investigate on the library side. I also saw some comments here that point in the same direction: https://github.com/zephyrproject-rtos/zephyr/pull/39210
Kind reagards

**Comment by rlubos** [comment]
Thank you @mglettig for the report, I'll try to investigate this further, but please open a bug report issue so we can keep track of its progress.

**Comment by mglettig** [comment]
> rsp->body_start

Hi @rlubos. I created the bug report here: https://github.com/zephyrproject-rtos/zephyr/issues/41683.

**Comment by mglettig** [comment]
> Thank you @mglettig for the report, I'll try to investigate this further, but please open a bug report issue so we can keep track of its progress.

Hi @rlubos. I created the bug report here: https://github.com/zephyrproject-rtos/zephyr/issues/41683

**Pull request merged**